### PR TITLE
Warn Windows 7 users about old TLS

### PR DESCRIPTION
**Pull request by matklad**
An eyepatch for https://github.com/rust-lang/cargo/issues/5066.

@retep998 what would the best way to check for windows 7 specifically? Currently I use just `#[cfg(windows)]`, which might give false positives on 8 and 10.

### CODE REVIEW

**Comment by sfackler** [comment]
This seems like it'd be helpful: https://msdn.microsoft.com/en-us/library/windows/desktop/dn424961(v=vs.85).aspx.

**Comment by Eh2406** [comment]
According to,
https://support.microsoft.com/en-us/help/3140245/update-to-enable-tls-1-1-and-tls-1-2-as-a-default-secure-protocols-in
Looks like the patch for windows requires or set a `DefaultSecureProtocols` registry entry. Maybe we can check for that?

And thanks for making this patch!

**Comment by matklad** [comment]
Amusingly, trying to test out this fix on win7 with 

```
cargo install --git https://github.com/matklad/cargo --branch win7-y-u-no-tls
```

fails with this very error =P

**Comment by matklad** [comment]
Ok, I've checked that the current patch works as expected on windows 7. 

I've looked into detecting registry key/windows version a bit, but I've failed to find a smiple API in winapi-rs to do this, and I don't feel confident about using complex windows API, so it would be great if someone with more windows experience implements a version check =)

**Comment by Arnavion** [comment]
- The KB article declares that the issue is for Windows Server 2012 which is the server equivalent of Windows 8, but doesn't list Windows 8 itself as affected. But this is probably because Windows 8 became unsupported some time before that update was released, so it probably is affected and including it in the check should not be a problem.

- The reg key doesn't exist on Windows 10 and Windows 10 works without it, so just testing the reg key's existence independent of the OS version will not be correct.

So  `IsWindows8Point10OrGreater` returning `FALSE` is probably all you need.

**Comment by Eh2406** [comment]
Unfortunately `IsWindows8Point10OrGreater ` is not yet in winapi-rs. :-(

I tried to use GetVersionEx directly:
```
    use winapi::um::winnt::OSVERSIONINFOW;
    use winapi::um::sysinfoapi::GetVersionExW;
    let mut os = OSVERSIONINFOW {
        dwOSVersionInfoSize: std::mem::size_of::<OSVERSIONINFOW>() as u32,
        dwMajorVersion: 0,
        dwMinorVersion: 0,
        dwBuildNumber: 0,
        dwPlatformId: 0,
        szCSDVersion: [0; 128],
    };
    unsafe { GetVersionExW(&mut os) };
    println!("{:?}.{:?}", os.dwMajorVersion, os.dwMinorVersion);
```
and on my win10, I get 6.2. according to the docs at https://msdn.microsoft.com/en-us/library/windows/desktop/ms724834(v=vs.85).aspx that is win8, probably something about not having a manifest file. But if we get 6.1 on win 7 we may have a workaround. (I don't have a win7 box to test on.)

**Comment by Arnavion** [comment]
@Eh2406 Yes, as I said in my comment, the function is header-only, and lies to applications without manifests.

**Comment by Arnavion** [comment]
Something like this will work, once you've figured out the manifest problem (not tested or compiled):

```rust
const WIN32_WINNT_WINBLUE_HIBYTE = 0x06;
const WIN32_WINNT_WINBLUE_LOBYTE = 0x06;

fn IsWindows8Point1OrGreater() -> bool
{
    IsWindowsVersionOrGreater(WIN32_WINNT_WINBLUE), WIN32_WINNT_WINBLUE_LOBYTE, 0)
}

fn IsWindowsVersionOrGreater(majorVersion: DWORD, minorVersion: DWORD, servicePackMajor: DWORD) -> bool {
    let conditionMask: DWORDLONG = VerSetConditionMask(VerSetConditionMask(VerSetConditionMask(0, VER_MAJORVERSION, VER_GREATER_EQUAL), VER_MINORVERSION, VER_GREATER_EQUAL), VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL);

   let mut osvi: OSVERSIONINFOEXW = std::mem::zeroed();
   osvi.dwMajorVersion = majorVersion;
   osvi.dwMinorVersion = minorVersion;
   osvi.wServicePackMajor = servicePackMajor;

  VerifyVersionInfoW(&mut osvi, VER_MAJORVERSION | VER_MINORVERSION | VER_SERVICEPACKMAJOR, dwlConditionMask) != FALSE
}
```

**Comment by Eh2406** [comment]
@Arnavion, I can't find your comment about `manifests`. I am sorry if I echoed info you had already provided.

> once you've figured out the manifest problem

I don't think the 'manifest problem' is trivial. I'd love to be proven wrong. I don't know of a way to add a manifest that doesn't involve a large build dependency. The win SDK for `mt.exe` or LLVM for its alternative. Do we want to require that for building cargo?

> not tested or compiled

Yes that is a better solution, once it is converted to valid rust.

**Comment by Arnavion** [comment]
@retep998 should comment about that. It's not a problem for windows-msvc builds of course, but I don't know anything about windows-gnu (assuming the cargo.exe that ships with the windows-gnu toolchain is itself built with windows-gnu and not just the same binary as the one that ships with the windows-msvc toolchain).

**Comment by Eh2406** [comment]
If we need to deal with the manifest to lowerer the false positive rate, then perhaps we merge as is for now so it can get in `1.24.1`. And add a `FIXME` to get dealt with wen manifests are fixed.

File: src/cargo/util/network.rs [hunk 1]
```diff
@@ -33,6 +33,30 @@ fn maybe_spurious(err: &Error) -> bool {
...
+}
+
+#[cfg(not(windows))]
+fn should_warn_about_old_tls_for_win7(_err: &Error) -> bool {
+    false
+}
+
+const WIN7_TLS_WARNING: &str = "\
+Certificate check failure might be caused by outdated TLS on Windows 7.\n\
+Please follow these instructions to enable more secure TLS:\n\n    \
```
Comment (retep998) [inline src/cargo/util/network.rs]: Until we are able to reliably detect Windows 7, perhaps we should prefix this with `If you are using Windows 7, please...`

**Comment by Arnavion** [comment]
Agreed. The function's comment and error message should also mention Server 2008R2 and Server 2012 in addition to 7, and perhaps word it like "If you're on ... then certificate check failure might be caused..."

**Comment by retep998** [comment]
Having the necessary tooling from VC++/MinGW to link manifests is not an issue. The issue is having the support from Rust itself. It will involve non-trivial design work to come up with a good way to link manifests, and I doubt it will happen any time soon.

However there is a shorter term solution for manifests and that is having simple side by side manifests instead of embedded manifests. It would involve someone writing manifests such as `cargo.manifest` and then getting the Rust packaging stuff to shove that in the same component and folder as `cargo.exe`.

**Comment by Eh2406** [comment]
I think we can use [rust-embed-resource](https://github.com/nabijaczleweli/rust-embed-resource) or similar build script magic to include the manifest, without the design work from rustic. (If we have the tooling from VC++/MinGW installed.)

I have never had any luck getting `side by side manifests` to work, but it does sound like a simple solution.

**Comment by matklad** [comment]
Hm, it looks like we won't be able to implement the manifest solution in any reasonable time frame :-( 

Could we check for the presence of `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\WinHttp` instead, as suggested by @Eh2406?

Specifically, is this key present on Windows 8 and Windows 10?

**Comment by matklad** [comment]
Updated the error message to mention windows server and to be conditional on "if you are indeed on windows". 

It seems unlikely to me that we will be able to implement the `cargo.exe.manifest` solution fast, and this I think means that this patch is the best we can hope for 1.24.1. 

The failure mode here is that a user on Windows 10 might get this wrong worming if threre's genuinely something wrong with TLS. However, because we are checking for a specific code from libgit, this seams unlikely?

**Comment by alexcrichton** [comment]
Thanks for the quick fix here @matklad! It sounds like this is hopefully going to be a temporary holdover until https://github.com/libgit2/libgit2/pull/4550 is sorted out, after which we'll probably just update libgit2 pronto and get that into Cargo. In that sense it seems like we'll be using this for hopefully only a few cycles, but still a few cycles in the sense that it'll get backported and such.

This looks pretty conservative but I think one final thing we may wish to do is to only warn about connections to github.com perhaps? I believe that's the main instigator for this issue, right? It may also help reduce the noise of warnings in legitimate situations as that seems to come up relatively often (certificate errors on Windows).

I'd personally also be much more comfortable backporting a warning like this as well rather than libgit2 updates, so I think for sure we should pursue this.

@matklad would you be ok switching to only warning on github.com hostnames?

**Comment by matklad** [comment]
>@matklad would you be ok switching to only warning on github.com hostnames?

Sure will try to do this in a couple of hours!

**Comment by matklad** [comment]
Updated to check for GitHub specifically, and verified on the virtual machine that the message is indeed printed on an unpatched win7, and that a patched one works OK.

**Pull request merged**